src/main/scala/format/pgn/Parser.scala

```diff
@@ -98,296 +63,200 @@ object Parser {
 
-    def glyphs: Parser[Glyphs] =
-      as("glyphs") {
-        rep(glyph) ^^ Glyphs.fromList
-      }
+    val promotion: P[PromotableRole] = P.char('=').?.with1 *> mapParser(promotable, "promotion")
 
-    def glyph: Parser[Glyph] =
-      as("glyph") {
-        mapParser(
-          Glyph.MoveAssessment.all.sortBy(_.symbol.length).map { g =>
-            g.symbol -> g
-          },
-          "glyph"
-        )
-      }
+    // e5
+    val pawn: P[Std] = dest.map(Std(_, Pawn))
 
-    val x = exists("x")
+    // Bg5
+    val ambigous: P[Std] = (role ~ x ~ dest).map { case ((ro, ca), de) =>
+      Std(dest = de, role = ro, capture = ca)
+    }
 
-    val check = exists("+")
+    // B@g5
+    val drop: P[Drop] = ((role <* P.char('@')) ~ dest).map { case (role, pos) => Drop(role, pos) }
 
-    val checkmate = ("#" | "++") ^^^ true | success(false)
+    val pawnDrop: P[Drop] = (P.char('@') *> dest).map(Drop(Pawn, _))
 
-    val role = mapParser(Role.allByPgn, "role") | success(Pawn)
+    // Bac3 Baxc3 B2c3 B2xc3 Ba2xc3
+    val disambiguated: P[Std] = (role ~ file.? ~ rank.? ~ x ~ dest).map { case ((((ro, fi), ra), ca), de) =>
+      Std(dest = de, role = ro, capture = ca, file = fi, rank = ra)
+    }
 
-    val file = mapParser(fileMap, "file")
+    // d7d5
+    val disambiguatedPawn: P[Std] = (((file.? ~ rank.?) ~ x).with1 ~ dest).map { case (((fi, ra), ca), de) =>
+      Std(dest = de, role = Pawn, capture = ca, file = fi, rank = ra)
+    }
 
-    val rank = mapParser(rankMap, "rank")
+    val suffixes: P0[Suffixes] = (promotion.? ~ checkmate ~ check ~ glyphs).map { case (((p, cm), c), g) =>
+      Suffixes(c, cm, p, g)
+    }
 
-    val promotable = Role.allPromotableByPgn mapKeys (_.toUpper)
+    val castle: P[San] = (qCastle | kCastle).map(Castle(_))
 
-    val promotion = ("=" ?) ~> mapParser(promotable, "promotion")
+    val standard: P[San] =
+      disambiguatedPawn.backtrack | pawn.backtrack | disambiguated.backtrack | ambigous.backtrack | drop.backtrack | pawnDrop.backtrack
 
-    val dest = mapParser(Pos.allKeys, "dest")
+    val move: P[San] = (castle | standard).withContext("Invalid chess move")
+    val moveWithSuffix: P[San] = (move ~ suffixes <* whitespaces)
+      .map { case (std, suf) =>
+        std withSuffixes suf
+      }
 
-    def exists(c: String): Parser[Boolean] = c ^^^ true | success(false)
+    def apply(str: String): Validated[String, San] =
+      moveWithSuffix.parse(str) match {
+        case Right((_, san)) => valid(san)
+        case Left(err)       => invalid(showExpectations("Cannot parse move", str, err))
+      }
 
-    def mapParser[A, B](pairs: Iterable[(A, B)], name: String): Parser[B] =
-      pairs.foldLeft(failure(name + " not found"): Parser[B]) { case (acc, (a, b)) =>
-        a.toString ^^^ b | acc
+    def mapParser[A, B](pairs: Iterable[(A, B)], name: String): P[B] =
+      pairs.foldLeft(P.failWith(name + " not found"): P[B]) { case (acc, (a, b)) =>
+        P.string(a.toString).as(b) | acc
```

Comment:
nit: if it were me, I would make this take `Iterable[(String, B)]` rather than call toString, since it is more explicit that it is just a string parser in that case.

Also, if this list is long, `stringIn` would be faster (supposedly, it does use a tree, but for small lists of strings it probably won't be, so you might want to experiment).

An alternative would be:

```scala
val pairMap = pairs.view.map { case (a, b) => (a.toString, b) }.to(Map)
P.stringIn(pairMap.keySet).map(pairMap(_)) | P.failWith(name + " not found")
```

Reply:
Great, I'll try your options to see which one is better.